File: packages/react-native-nitro-image-blurhash/android/src/main/java/com/margelo/nitro/image/blurhash/HybridBlurHashFactory.kt
package com.margelo.nitro.image.blurhash

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import com.margelo.nitro.image.HybridImage
import com.margelo.nitro.image.HybridImageSpec
import com.margelo.nitro.image.extensions.toCpuAccessible
import com.mrousavy.blurhash.BlurHashDecoder
import com.mrousavy.blurhash.BlurHashEncoder

@DoNotStrip
@Keep
class HybridBlurHashFactory : HybridBlurHashFactorySpec() {
    override fun encode(image: HybridImageSpec, componentsX: Double, componentsY: Double): String {
        val native = image as? HybridImage
            ?: throw Error("The given image ($image) is not a `HybridImage` and cannot be encoded to a BlurHash!")
        return BlurHashEncoder.encode(native.bitmap.toCpuAccessible(), componentsX.toInt(), componentsY.toInt())
    }

    override fun encodeAsync(image: HybridImageSpec, componentsX: Double, componentsY: Double): Promise<String> {
        return Promise.async { encode(image, componentsX, componentsY) }
    }

    override fun decode(blurhash: String, width: Double, height: Double, punch: Double): HybridImageSpec {
        val bitmap = BlurHashDecoder.decode(blurhash, width.toInt(), height.toInt(), punch.toFloat())
            ?: throw Error("The given BlurHash ($blurhash) is not a valid BlurHash!")
        return HybridImage(bitmap)
    }

    override fun decodeAsync(blurhash: String, width: Double, height: Double, punch: Double): Promise<HybridImageSpec> {
        return Promise.async { decode(blurhash, width, height, punch) }
    }

    override fun clearCosineCache() {
        BlurHashDecoder.clearCache()
    }
}

File: packages/react-native-nitro-image-blurhash/ios/HybridBlurHashFactory.swift
import Foundation
import UIKit
import NitroModules
import NitroImage

fileprivate class HybridImage: HybridImageSpec, NativeImage {
  let uiImage: UIImage
  init(uiImage: UIImage) {
    self.uiImage = uiImage
    super.init()
  }
}

class HybridBlurHashFactory: HybridBlurHashFactorySpec {
  func encode(image: any HybridImageSpec, componentsX: Double, componentsY: Double) throws -> String {
    guard let native = image as? NativeImage else {
      throw RuntimeError.error(withMessage: "The given image (\(image)) is not a `NativeImage` and cannot be encoded to a BlurHash!")
    }
    guard let hash = native.uiImage.encodeBlurhash(numberOfComponents: (Int(componentsX), Int(componentsY))) else {
      throw RuntimeError.error(withMessage: "Failed to encode BlurHash!")
    }
    return hash
  }

  func encodeAsync(image: any HybridImageSpec, componentsX: Double, componentsY: Double) throws -> Promise<String> {
    return Promise.async {
      return try self.encode(image: image, componentsX: componentsX, componentsY: componentsY)
    }
  }

  func decode(blurhash: String, width: Double, height: Double, punch: Double) throws -> any HybridImageSpec {
    let size = CGSize(width: width, height: height)
    guard let uiImage = UIImage(blurHash: blurhash, size: size, punch: Float(punch)) else {
      throw RuntimeError.error(withMessage: "The given BlurHash (\(blurhash)) is not a valid BlurHash!")
    }
    return HybridImage(uiImage: uiImage)
  }

  func decodeAsync(blurhash: String, width: Double, height: Double, punch: Double) throws -> Promise<any HybridImageSpec> {
    return Promise.async {
      return try self.decode(blurhash: blurhash, width: width, height: height, punch: punch)
    }
  }

  func clearCosineCache() throws {
    // No-op: the iOS BlurHash decoder has no cosine cache.
  }
}
